File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/repository/PreferencesRepository.kt
package com.bissbilanz.repository

import app.cash.sqldelight.coroutines.asFlow
import app.cash.sqldelight.coroutines.mapToOneOrNull
import com.bissbilanz.api.BissbilanzApi
import com.bissbilanz.api.generated.model.Preferences
import com.bissbilanz.api.generated.model.PreferencesUpdate
import com.bissbilanz.cache.BissbilanzDatabase
import com.bissbilanz.mode.AppModeManager
import com.bissbilanz.sync.SyncOperation
import com.bissbilanz.sync.SyncQueue
import com.bissbilanz.userdata.UserDataDatabase
import com.bissbilanz.util.decodeOrNull
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlin.time.Clock

class PreferencesRepository(
    private val api: BissbilanzApi,
    private val db: UserDataDatabase,
    private val cacheDb: BissbilanzDatabase,
    private val syncQueue: SyncQueue,
    private val json: Json,
    private val appModeManager: AppModeManager,
) {
    fun preferences(): Flow<Preferences?> =
        db.userDataDatabaseQueries
            .selectPreferences()
            .asFlow()
            .mapToOneOrNull(Dispatchers.IO)
            .map { cached ->
                cached?.let {
                    json.decodeOrNull<Preferences>(it.jsonData) ?: run {
                        db.userDataDatabaseQueries.deletePreferences()
                        null
                    }
                }
            }

    suspend fun refresh() {
        if (appModeManager.isLocal) return
        val prefs = api.getPreferences()
        cachePreferences(prefs)
    }

    /**
     * Reports the device's IANA timezone to the server so server-side analytics/MCP
     * bucket days/hours in the user's local tz. Only updates when it differs from the
     * stored value (loop guard); compares against the authoritative server value.
     */
    suspend fun reportTimeZone(deviceTimeZone: String) {
        if (appModeManager.isLocal) return
        val current = api.getPreferences()
        if (current.timeZone == deviceTimeZone) return
        updatePreferences(PreferencesUpdate(timeZone = deviceTimeZone))
    }

    suspend fun updatePreferences(update: PreferencesUpdate): Preferences {
        val cached = db.userDataDatabaseQueries.selectPreferences().executeAsOneOrNull()
        val current =
            cached?.let {
                json.decodeOrNull<Preferences>(it.jsonData) ?: run {
                    db.userDataDatabaseQueries.deletePreferences()
                    null
                }
            } ?: Preferences(
                showChartWidget = true,
                showFavoritesWidget = true,
                showSupplementsWidget = true,
                showWeightWidget = true,
                showMealBreakdownWidget = true,
                showTopFoodsWidget = true,
                showSleepWidget = true,
                widgetOrder = emptyList(),
                mealOrder = emptyList(),
                startPage = "dashboard",
                favoriteTapAction = "instant",
                favoriteMealAssignmentMode = "time_based",
                favoriteMealTimeframes = emptyList(),
                visibleNutrients = emptyList(),
                locale = null,
                timeZone = "UTC",
            )
        val updated = applyUpdate(current, update)
        cachePreferences(updated)
        syncQueue.enqueue(SyncOperation.UpdatePreferences(json.encodeToString(update)))
        return updated
    }

    private fun cachePreferences(prefs: Preferences) {
        db.userDataDatabaseQueries.insertPreferences(
            jsonData = json.encodeToString(prefs),
        )
        // SyncMeta lives in the cache database; written after the user-data write.
        cacheDb.bissbilanzDatabaseQueries.upsertSyncMeta(
            entityType = "preferences",
            lastSyncedAt = Clock.System.now().toString(),
        )
    }

    private fun applyUpdate(
        current: Preferences,
        update: PreferencesUpdate,
    ): Preferences =
        current.copy(
            showChartWidget = update.showChartWidget ?: current.showChartWidget,
            showFavoritesWidget = update.showFavoritesWidget ?: current.showFavoritesWidget,
            showSupplementsWidget = update.showSupplementsWidget ?: current.showSupplementsWidget,
            showWeightWidget = update.showWeightWidget ?: current.showWeightWidget,
            showMealBreakdownWidget = update.showMealBreakdownWidget ?: current.showMealBreakdownWidget,
            showTopFoodsWidget = update.showTopFoodsWidget ?: current.showTopFoodsWidget,
            showSleepWidget = update.showSleepWidget ?: current.showSleepWidget,
            widgetOrder = update.widgetOrder?.map { it.value } ?: current.widgetOrder,
            mealOrder = update.mealOrder ?: current.mealOrder,
            startPage = update.startPage?.value ?: current.startPage,
            favoriteTapAction = update.favoriteTapAction?.value ?: current.favoriteTapAction,
            favoriteMealAssignmentMode = update.favoriteMealAssignmentMode?.value ?: current.favoriteMealAssignmentMode,
            favoriteMealTimeframes = current.favoriteMealTimeframes,
            visibleNutrients = update.visibleNutrients ?: current.visibleNutrients,
            locale = update.locale?.value ?: current.locale,
            timeZone = update.timeZone ?: current.timeZone,
        )
}

File: mobile/iosApp/Bissbilanz/Persistence/PreferencesRepository.swift
import Foundation
import Observation
import SwiftData

/// Local-first repository for user preferences (singleton row). Updates merge
/// the partial `PreferencesUpdate` onto the cached value locally and queue
/// the patch via the sync manager, mirroring the Android repository.
@MainActor
@Observable
final class PreferencesRepository {
    private let context: ModelContext
    private let api: BissbilanzAPI
    private let appMode: AppModeManager
    private let syncManager: SyncManager

    init(context: ModelContext, api: BissbilanzAPI, appMode: AppModeManager, syncManager: SyncManager) {
        self.context = context
        self.api = api
        self.appMode = appMode
        self.syncManager = syncManager
    }

    func preferences() -> Preferences? {
        fetchRow()?.toPreferences()
    }

    func refresh() async throws {
        guard !appMode.isLocal else { return }
        let prefs = try await api.getPreferences()
        // Singleton row — a queued updatePreferences carries no affectedId, so
        // presence is the only guard available (see GoalsRepository.refresh).
        guard !syncManager.hasPending(table: "preferences") else { return }
        upsert(prefs)
        save()
    }

    /// Reports the device's IANA timezone to the server so server-side analytics/MCP
    /// bucket days/hours in the user's local tz. Only updates when it differs from the
    /// stored value (loop guard); compares against the authoritative server value.
    func reportTimeZone(_ deviceTimeZone: String) async throws {
        guard !appMode.isLocal else { return }
        let current = try await api.getPreferences()
        guard current.timeZone != deviceTimeZone else { return }
        _ = try await update(PreferencesUpdate(timeZone: deviceTimeZone))
    }

    @discardableResult
    func update(_ update: PreferencesUpdate) async throws -> Preferences {
        let current = preferences() ?? .defaults
        let patch = (try? JSONPatch.dictionary(of: update)) ?? [:]
        let merged = (try? JSONPatch.merged(Preferences.self, base: current, patch: patch)) ?? current
        upsert(merged)
        save()
        syncManager.enqueue(.updatePreferences(body: update))
        return merged
    }

    // MARK: - Store helpers

    private func fetchRow() -> LocalPreferences? {
        let singletonId = LocalPreferences.singletonId
        var descriptor = FetchDescriptor<LocalPreferences>(predicate: #Predicate { $0.id == singletonId })
        descriptor.fetchLimit = 1
        return (try? context.fetch(descriptor))?.first
    }

    private func upsert(_ preferences: Preferences) {
        if let row = fetchRow() {
            row.update(from: preferences)
        } else {
            context.insert(LocalPreferences(preferences: preferences))
        }
    }

    private func save() {
        try? context.save()
    }
}
